### PR TITLE
FEATURE: use new max v3 api to query closed orders by timestamp

### PR DESCRIPTION
**Pull request by kbearXD**

### CODE REVIEW

File: pkg/exchange/max/exchange.go [hunk 1]
```diff
@@ -311,6 +315,47 @@ func (e *Exchange) queryClosedOrdersByLastOrderID(ctx context.Context, symbol st
...
+	market := toLocalSymbol(symbol)
+	walletType := maxapi.WalletTypeSpot
+	if e.MarginSettings.IsMargin {
+		walletType = maxapi.WalletTypeMargin
+	}
+
+	req := e.v3client.NewGetWalletClosedOrdersRequest(walletType).Market(market).Limit(1000)
+
+	if !until.IsZero() {
+		req.Timestamp(until)
```
Comment (c9s) [inline pkg/exchange/max/exchange.go:332]: what will happen if we don't set Timestamp?

File: pkg/exchange/max/exchange.go [hunk 2]
```diff
@@ -311,6 +315,47 @@ func (e *Exchange) queryClosedOrdersByLastOrderID(ctx context.Context, symbol st
...
+	}
+
+	maxOrders, err := req.Do(ctx)
```
Comment (c9s) [inline pkg/exchange/max/exchange.go:335]: and, what's the default ordering?

File: pkg/exchange/max/exchange.go [hunk 3]
```diff
@@ -311,6 +315,47 @@ func (e *Exchange) queryClosedOrdersByLastOrderID(ctx context.Context, symbol st
...
+	if err != nil {
+		return orders, err
+	}
+
+	for _, maxOrder := range maxOrders {
+		if maxOrder.CreatedAt.Time().Before(since) {
+			break
```
Comment (c9s) [inline pkg/exchange/max/exchange.go:342]: break? maybe just "continue" and skip it?

File: pkg/exchange/max/maxapi/v3/get_wallet_closed_orders_request.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+//go:generate GetRequest -url "/api/v3/wallet/:walletType/orders/closed" -type GetWalletClosedOrdersRequest -responseType []Order
+type GetWalletClosedOrdersRequest struct {
+	client requestgen.AuthenticatedAPIClient
+
+	walletType WalletType `param:"walletType,slug,required"`
+
+	market    string     `param:"market,required"`
+	timestamp *time.Time `param:"timestamp,milliseconds,omitempty"`
+	orderBy   *string    `param:"order_by,omitempty" validValues:"asc,desc,asc_updated_at,desc_updated_at"`
```
Comment (c9s) [inline pkg/exchange/max/maxapi/v3/get_wallet_closed_orders_request.go:25]: can you define these strings as constants ? "asc,desc,asc_updated_at,desc_updated_at"

Comment (c9s) [reply pkg/exchange/max/maxapi/v3/get_wallet_closed_orders_request.go:25]: just like WalletType

File: pkg/exchange/max/exchange.go (hunk 1 above)
Comment (kbearXD) [reply pkg/exchange/max/exchange.go:332]: it will use now timestamp as default to query closed orders

File: pkg/exchange/max/exchange.go (hunk 2 above)
Comment (kbearXD) [reply pkg/exchange/max/exchange.go:335]: desc. there are four order type - asc, desc, asc_updated_at, desc_updated_at

Comment (c9s) [reply pkg/exchange/max/exchange.go:335]: but we should use asc here

Comment (kbearXD) [reply pkg/exchange/max/exchange.go:335]: ok!

**Review by c9s — APPROVED** [review]

File: pkg/exchange/max/exchange.go [hunk 5]
```diff
@@ -268,6 +268,10 @@ func (e *Exchange) QueryOpenOrders(ctx context.Context, symbol string) (orders [
 // lastOrderID is not supported on MAX
 func (e *Exchange) QueryClosedOrders(ctx context.Context, symbol string, since, until time.Time, lastOrderID uint64) ([]types.Order, error) {
 	log.Warn("!!!MAX EXCHANGE API NOTICE!!! the since/until conditions will not be effected on closed orders query, max exchange does not support time-range-based query")
+	if !since.IsZero() || !until.IsZero() {
```
Comment (c9s) [inline pkg/exchange/max/exchange.go:271]: should be until.IsZero()? since the api only support one timestamp parameter?

Comment (kbearXD) [reply pkg/exchange/max/exchange.go:271]: We can give one of it. If we give until, we will use 2018-01-01 as timestamp to query and filter it by until.

**Pull request merged**